src/xma/src/xmaapi/xmaadmin.cpp

```diff
@@ -247,16 +248,54 @@ xma_admin_session_destroy(XmaAdminSession *session)
     /*
     delete (XmaHwSessionPrivate*)session->base.hw_session.private_do_not_use;
     */
+    XmaHwSessionPrivate *priv1 = (XmaHwSessionPrivate*)session->base.hw_session.private_do_not_use;
+    if(priv1 != nullptr) {
+        priv1->CU_cmds.clear();
+        priv1->CU_error_cmds.clear();
+        priv1->execbo_lru.clear();
+        priv1->execbo_to_check.clear();
+        for(auto& kernel_bo: priv1->kernel_execbos)
+        {
+            xclUnmapBO(priv1->dev_handle, kernel_bo.handle, kernel_bo.data);
+            xclFreeBO(priv1->dev_handle, kernel_bo.handle);
+        }
+        priv1->kernel_execbos.clear();
+        priv1->buffer_pools.clear();
+        priv1->dev_handle = NULL;
+        priv1->kernel_info = NULL;
+        priv1->device = NULL;
+        delete priv1;
+    }
+
     session->base.hw_session.private_do_not_use = nullptr;
     session->base.plugin_data = nullptr;
     session->base.stats = NULL;
     session->admin_plugin = NULL;
     //do not change kernel in_use as it maybe in use by another plugin
     session->base.hw_session.dev_index = -1;
     session->base.session_signature = NULL;
+    session->base.channel_id = -1;
     free(session);
     session = nullptr;
 
+    //Free session from all_sessions_vec
+    int vec_size = g_xma_singleton->all_sessions_vec.size();
```

Comment:
create a single function and call it from multiple places instead of repeating the code.